hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/impl/GetS3RegionOperation.java

```diff
@@ -0,0 +1,140 @@
+import org.apache.hadoop.fs.store.LogExactlyOnce;
+import org.apache.hadoop.util.DurationInfo;
+
+import static org.apache.hadoop.fs.s3a.Constants.BUCKET_REGION_HEADER;
+import static org.apache.hadoop.fs.s3a.Constants.FS_S3A_BUCKET_PREFIX;
+import static org.apache.hadoop.fs.s3a.Statistic.STORE_REGION_PROBE;
+import static org.apache.hadoop.fs.s3a.impl.InternalConstants.SC_301_MOVED_PERMANENTLY;
+import static org.apache.hadoop.fs.s3a.impl.InternalConstants.SC_404_NOT_FOUND;
+import static org.apache.hadoop.fs.statistics.impl.IOStatisticsBinding.trackDuration;
+
+/**
+ * Operation to get the S3 region for a bucket.
+ * If a region is not specified in {@link org.apache.hadoop.fs.s3a.Constants#AWS_REGION}, this
+ * operation will warn and then probe S3 to get the region.
+ */
+public class GetS3RegionOperation extends ExecutingStoreOperation<Region> {
+
+  private final AWSCredentialProviderList credentials;
+
+  private final String region;
+
+  private final String bucket;
+
+  private final StoreContext context;
+
+  private final static Map<String, Region> BUCKET_REGIONS = new HashMap<>();
+
+  private static final Logger LOG = LoggerFactory.getLogger(
+      GetS3RegionOperation.class);
+
+  /** Exactly once log to warn about setting the region in config to avoid probe. */
+  private static final LogExactlyOnce SET_REGION_WARNING = new LogExactlyOnce(LOG);
+
+  public GetS3RegionOperation(
+      final StoreContext context,
+      String region,
+      String bucket,
+      AWSCredentialProviderList credentials) {
+      super(context);
+      this.context = context;
+      this.credentials = credentials;
+      this.region = region;
+      this.bucket = bucket;
+  }
+
+
+  @Override
+  public Region execute() throws IOException {
+    
+    if (!StringUtils.isBlank(region)) {
+      return Region.of(region);
+    }
+
+    Region cachedRegion = BUCKET_REGIONS.get(bucket);
+
+    if (cachedRegion != null) {
+      LOG.debug("Got region {} for bucket {} from cache", cachedRegion, bucket);
+      return cachedRegion;
+    }
+
+    try (DurationInfo ignore =
+        new DurationInfo(LOG, false, "Get S3 region")) {
+
+      SET_REGION_WARNING.warn(
+          "Getting region for bucket {} from S3, this will slow down FS initialisation. "
+              + "To avoid this, set the region using property {}", bucket,
+          FS_S3A_BUCKET_PREFIX + bucket + ".endpoint.region");
+
+      // build a s3 client with region eu-west-1 that can be used to get the region of the
+      // bucket. Using eu-west-1, as headBucket() doesn't work with us-east-1. This is because
+      // us-east-1 uses the endpoint s3.amazonaws.com, which resolves bucket.s3.amazonaws.com
+      // to the actual region the bucket is in. As the request is signed with us-east-1 and
+      // not the bucket's region, it fails.
+      S3Client getRegionS3Client =
+          S3Client.builder().region(Region.EU_WEST_1).credentialsProvider(credentials)
```

Comment:
what's going to happen with govcloud or china regions?

I'd propose having a callback interface we've done with the other operations, so this class can be given unit tests with a different implementation.